cincinnati/src/plugins/mod.rs

```diff
@@ -0,0 +1,285 @@
+    }
+}
+
+/// Convert from an IntenralIO to a PluginExchange
+///
+/// This conversion cannot fail.
+impl From<InternalIO> for interface::PluginExchange {
+    fn from(internal_io: InternalIO) -> Self {
+        let mut plugin_exchange = Self::new();
+
+        plugin_exchange.set_graph(internal_io.graph.into());
+        plugin_exchange.set_parameters(internal_io.parameters);
+
+        plugin_exchange
+    }
+}
+
+/// Convert from InternalIO to ExternalIO
+///
+/// The serialization can apparently fail.
+impl TryFrom<InternalIO> for ExternalIO {
+    type Err = Error;
+
+    fn try_from(internal_io: InternalIO) -> Fallible<Self> {
+        use protobuf::Message;
+        let plugin_exchange: interface::PluginExchange = internal_io.into();
+        let bytes: Vec<u8> = plugin_exchange.write_to_bytes()?;
+        Ok(Self { bytes })
+    }
+}
+
+/// Try to convert from PluginIO to InternalIO
+///
+/// In case of the variant Plugion::ExternalIO this involves deserialization
+/// which might fail.
+impl TryFrom<PluginIO> for InternalIO {
+    type Err = Error;
+
+    fn try_from(plugin_io: PluginIO) -> Fallible<Self> {
+        match plugin_io {
+            PluginIO::InternalIO(internal_io) => Ok(internal_io),
+            PluginIO::ExternalIO(external_io) => external_io.try_into(),
+        }
+    }
+}
+
+/// Copmatibility impl to automatically convert the unvarianted PluginIO when the
+/// variant struct ExternalIO is expented.
+///
+/// This may fail due to the possible failing conversion.
+impl TryFrom<PluginIO> for ExternalIO {
+    type Err = Error;
+
+    fn try_from(plugin_io: PluginIO) -> Fallible<Self> {
+        let final_io = match plugin_io {
+            PluginIO::ExternalIO(external_io) => external_io,
+            PluginIO::InternalIO(internal_io) => internal_io.try_into()?,
+        };
+        Ok(final_io)
+    }
+}
+
+/// Wrapper struct for the implementation for implementing Plugin<PluginIO>
+pub struct InternalPluginWrapper<T>(pub T);
+
+/// Wrapper struct for the implementation for implementing Plugin<PluginIO>
+pub struct ExternalPluginWrapper<T>(pub T);
+
+impl<T> Plugin<PluginIO> for InternalPluginWrapper<T>
+where
+    T: InternalPlugin,
+{
+    fn run(&self, plugin_io: PluginIO) -> Fallible<PluginIO> {
+        println!("running internal plugin");
+        let internal_io = self.0.run_internal(plugin_io.try_into()?)?;
+        internal_io.try_into()
+    }
+}
+
+impl<T> Plugin<PluginIO> for ExternalPluginWrapper<T>
+where
+    T: ExternalPlugin,
+{
+    fn run(&self, plugin_io: PluginIO) -> Fallible<PluginIO> {
+        println!("running external plugin");
+        let external_io = self.0.run_external(plugin_io.try_into()?)?;
+        external_io.try_into()
+    }
+}
+
+/// Processes all given Plugins sequentially.InternalIO
+///
+/// This function automatically converts between the different IO representations
+/// if necessary.
+pub fn process<T: 'static>(
+    plugins: &[Box<T>],
+    initial_io: InternalIO,
```

Comment:
Cosmetic/Opinion: I tend to think as this function being the external entrypoint, thus taking a graph+params and not the internal type. It doesn't effectively change anything except for my code-skimming approach, so feel free to disregard the comment if you think otherwise.

Reply:
In my mental model this is an internal entrypoint, since it's called internally by the components main loop. Eventually it won't need to take `InternalIO` which includes a `Graph` and the parameters `HashMap`, as the first plugin will produce the first `Graph`. In that case we can replace `InternalIO` here with just the `HashMap` which is provided by the client and/or component configuration.

Reply:
Ack, let's defer this to the "eventually" point.